### ISSUE TITLE
Feature Request: Share a link which gets automatically invalid after x hours / days

### ISSUE THREAD

**Issue by pgrm** [sandstorm-io/sandstorm#1480]
I came to this idea, when a friend send me a glympse - http://glympse.com/

A link, where I can follow his location for the next 4h, afterwards the link is inactive. Doing something like that in sandstorm would be cool. You could keep your data to yourself, but at the moment you'd need to go and invalidate the shared links manually.

Another example I have is when we had an intern in the summer, it would be so cool if I could share all the documents with him just for the next month, without going through the apps later on checking where he has still access and removing him.

Or is this already somewhere mentioned under a different name? Couldn't find it.

**Comment by zarvox** [comment]
+1 to the concept of having sharing links with expiry dates.  The `ApiTokens` table already has a field `expires`, though we don't appear to set that to nonnull anywhere yet (or indeed, actually enforce the expiry date, should one have been provided).

One potential challenge would be to make the UI for adding such link restrictions without complicating the UI in the common case, but that sounds feasible.

For cleanliness, I'd say the additional UI (once the user has indicated a desire for an expiring link) should look something like:

### Expires: 

* Never
* in 1 hour
* in 1 day
* in 1 week
* in 1 month
* Custom

And then if Custom is selected, you get to pick on which day it expires at 11:59PM.

**Comment by mitar** [comment]
Should this be implemented on the UI level or on capabilities level? (That capabilities itself have encoded expiration time?)

**Comment by paulproteus** [comment]
Given the above, it seems that this needs:

- UI decisions & UI work (note that since expiring links are uncommon, it will probably take some discussion to figure out how to enable that in the sharing UI without distracting people who don't need this feature - the next step is probably to create a mockup for discussion).

- Backend implementation work - as @zarvox said, the `ApiTokens` table (which is the implementation of Sandstorm permission grants)

- UI decisions & code decisions about how an app can trigger a sharing window with some expiry by default. Presumably that would be a flag to `startSharing` - see https://docs.sandstorm.io/en/latest/developing/path/#helping-the-user-share-access

**Comment by kentonv** [comment]
Note that there's another subtly different use case to consider: Create a sharing link that expires after a short time, but people who actually click on it within that short time get long-term access. This is useful for defending against a sharing link leaking long after it served its purpose.

**Comment by pgrm** [comment]
So one thing is an invite link which expires and the other is the actual
share which expires. In which case the share doesn't have to be a link. It
can be a normal share with a user as well or?

What do you think?

**Comment by paulproteus** [comment]
@kentonv Interesting that you draw this distinction. In my mind, I was always imagining that a short-term sharing link would enable long-term access once redeemed.

Are there people on this thread who wish for short-term sharing links that don't allow long-term access, once redeemed? Honest question.

**Comment by pgrm** [comment]
Yes, I am :) my original idea came from the app which let's you share your
location which can be really useful for few hours or an evening to know
where people are, and afterwards you don't want to share it anyone. Maybe
it would be easier to solve this on application level, but sandstorm just
thought me that I can trust sandstorm more than the apps running inside

**Comment by paulproteus** [comment]
Ah hah. Thanks for taking the time to spell that out for me!
